File: backend/src/main/java/cn/etarch/mao/harness/core/CompactionService.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;
import cn.etarch.mao.harness.llm.ChatResponse;
import cn.etarch.mao.harness.llm.ChatUsage;
import cn.etarch.mao.harness.llm.LlmAdapter;
import cn.etarch.mao.harness.llm.LlmModelConfig;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** 从即将发送的正常请求派生一次全量任务交接请求。 */
@Slf4j
@Service
@RequiredArgsConstructor
public class CompactionService {

    private static final Pattern HANDOFF_PATTERN = Pattern.compile("<handoff>(.*?)</handoff>", Pattern.DOTALL);

    private final LlmAdapter llmAdapter;
    private final TokenEstimator tokenEstimator;

    public SessionCompactionResult compactSession(
            Long sessionId,
            long expectedOldBoundary,
            List<PersistedChatMessage> messages,
            List<Long> snapshotMessageIds,
            ChatRequest normalRequest,
            LlmModelConfig modelConfig,
            CompactionConfig config,
            AgentEventListener listener,
            AtomicBoolean cancelFlag) {
        if (!config.isEnabled() || messages == null || messages.isEmpty() || normalRequest == null) {
            return null;
        }
        checkCancelled(cancelFlag);

        int normalRequestTokens = tokenEstimator.estimateRequestTokens(normalRequest);
        int effectiveWindow = CompactionConfig.resolveEffectiveContextWindow(modelConfig, config);
        if (normalRequestTokens < effectiveWindow * config.getTriggerRatio()) {
            return null;
        }

        long started = System.currentTimeMillis();
        ChatRequest compactionRequest = deriveRequest(normalRequest,
                buildHandoffInstruction(config.getMaxSummaryTokens()));
        int compactionRequestTokens = tokenEstimator.estimateRequestTokens(compactionRequest);
        if (compactionRequestTokens >= effectiveWindow) {
            throw new CompactionContextOverflowException(compactionRequestTokens, effectiveWindow);
        }

        if (listener != null) {
            listener.onCompactionStart("session", messages.size(), normalRequestTokens);
        }
        log.info("Session handoff compaction triggered: sessionId={}, messages={}, normalRequestTokens={}, "
                        + "compactionRequestTokens={}, effectiveWindow={}",
                sessionId, messages.size(), normalRequestTokens, compactionRequestTokens, effectiveWindow);

        try {
            ValidatedHandoff handoff = invokeAndValidate(compactionRequest, modelConfig, cancelFlag);
            if (handoff == null) {
                ChatRequest retryRequest = deriveRequest(compactionRequest, correctionInstruction());
                int retryTokens = tokenEstimator.estimateRequestTokens(retryRequest);
                if (retryTokens >= effectiveWindow) {
                    throw new CompactionContextOverflowException(retryTokens, effectiveWindow);
                }
                handoff = invokeAndValidate(retryRequest, modelConfig, cancelFlag);
            }
            if (handoff == null) {
                log.warn("Session handoff compaction failed semantic contract after one correction: sessionId={}", sessionId);
                if (listener != null) {
                    listener.onCompactionEnd("session", 0, 0, System.currentTimeMillis() - started);
                }
                return null;
            }

            SessionCompactionResult result = buildSafeResult(expectedOldBoundary, messages,
                    snapshotMessageIds, handoff, normalRequestTokens, started);
            if (result == null) {
                log.warn("Session handoff compaction rejected non-physical-prefix snapshot: sessionId={}, oldBoundary={}",
                        sessionId, expectedOldBoundary);
                if (listener != null) {
                    listener.onCompactionEnd("session", 0, 0, System.currentTimeMillis() - started);
                }
                return null;
            }
            log.info("Session handoff generated: sessionId={}, boundary={} -> {}, promptTokens={}, cachedTokens={}, "
                            + "completionTokens={}, durationMs={}",
                    sessionId, expectedOldBoundary, result.newLastCompactedMessageId(),
                    result.promptTokens(), result.cachedTokens(), result.completionTokens(), result.durationMs());
            return result;
        } catch (RuntimeException e) {
            if (listener != null) {
                listener.onCompactionEnd("session", 0, 0, System.currentTimeMillis() - started);
            }
            throw e;
        }
    }

    ChatRequest deriveRequest(ChatRequest source, String appendedUserContent) {
        List<ChatRequest.Message> messages = new ArrayList<>();
        if (source.getMessages() != null) {
            messages.addAll(source.getMessages());
        }
        messages.add(ChatRequest.Message.builder().role("user").content(appendedUserContent).build());
        return ChatRequest.builder()
                .messages(messages)
                .tools(source.getTools())
                .temperature(source.getTemperature())
                .stream(false)
                .reasoning(source.getReasoning())
                .audio(source.getAudio())
                .build();
    }

    private ValidatedHandoff invokeAndValidate(ChatRequest request, LlmModelConfig modelConfig,
                                                AtomicBoolean cancelFlag) {
        checkCancelled(cancelFlag);
        try {
            ChatResponse response = llmAdapter.chat(request, modelConfig, cancelFlag);
            checkCancelled(cancelFlag);
            if (response == null || response.getChoices() == null || response.getChoices().isEmpty()
                    || response.getChoices().get(0) == null
                    || response.getChoices().get(0).getMessage() == null) {
                log.warn("Compaction response is missing choices/message");
                return null;
            }
            ChatRequest.Message message = response.getChoices().get(0).getMessage();
            if (message.getToolCalls() != null && !message.getToolCalls().isEmpty()) {
                log.warn("Compaction response attempted {} tool call(s); ignored", message.getToolCalls().size());
                return null;
            }
            String content = TokenEstimator.contentToString(message.getContent());
            if (content == null) return null;
            Matcher matcher = HANDOFF_PATTERN.matcher(content);
            if (!matcher.find()) return null;
            String text = matcher.group(1).trim();
            if (text.isEmpty()) return null;
            return new ValidatedHandoff(text, response.getUsage());
        } catch (CompactionCancelledException e) {
            throw e;
        } catch (RuntimeException e) {
            if (isCancelled(cancelFlag) || (e.getMessage() != null && e.getMessage().contains("Cancelled by user"))) {
                throw new CompactionCancelledException(e);
            }
            throw e;
        }
    }

    private SessionCompactionResult buildSafeResult(long oldBoundary,
                                                     List<PersistedChatMessage> messages,
                                                     List<Long> snapshotMessageIds,
                                                     ValidatedHandoff handoff,
                                                     int beforeRequestTokens,
                                                     long started) {
        PersistedChatMessage last = messages.get(messages.size() - 1);
        long candidateBoundary = last.messageId();
        if (candidateBoundary <= oldBoundary || !isCompletePhysicalPrefix(
                oldBoundary, candidateBoundary, snapshotMessageIds, messages)) {
            return null;
        }
        int compactedCount = (int) snapshotMessageIds.stream()
                .filter(id -> id > oldBoundary && id <= candidateBoundary)
                .count();
        ChatUsage usage = handoff.usage();
        Integer cachedTokens = usage != null && usage.getPromptTokensDetails() != null
                ? usage.getPromptTokensDetails().getCachedTokens() : null;
        int promptTokens = usage != null ? usage.getPromptTokens() : 0;
        int completionTokens = usage != null ? usage.getCompletionTokens() : 0;
        int summaryTokens = tokenEstimator.estimateMessages(List.of(buildHandoffUserMessage(handoff.text())));
        return new SessionCompactionResult(
                handoff.text(), oldBoundary, candidateBoundary, last.persistedContentSnapshot(),
                compactedCount, promptTokens, cachedTokens, completionTokens,
                summaryTokens, 0, beforeRequestTokens, System.currentTimeMillis() - started);
    }

    private boolean isCompletePhysicalPrefix(long oldBoundary, long candidateBoundary,
                                             List<Long> snapshotMessageIds,
                                             List<PersistedChatMessage> messages) {
        if (snapshotMessageIds == null || snapshotMessageIds.isEmpty()) return false;
        Set<Long> normalizedIds = new HashSet<>();
        for (PersistedChatMessage message : messages) {
            normalizedIds.add(message.messageId());
        }
        return snapshotMessageIds.stream()
                .filter(id -> id > oldBoundary && id <= candidateBoundary)
                .allMatch(normalizedIds::contains)
                && snapshotMessageIds.contains(candidateBoundary);
    }

    public List<ChatRequest.Message> prependSessionSummary(
            String summary, List<ChatRequest.Message> incrementalMessages) {
        List<ChatRequest.Message> result = new ArrayList<>();
        if (summary != null && !summary.isBlank()) {
            result.add(buildHandoffUserMessage(summary));
        }
        if (incrementalMessages != null) {
            result.addAll(incrementalMessages);
        }
        return result;
    }

    ChatRequest.Message buildHandoffUserMessage(String summary) {
        return ChatRequest.Message.builder()
                .role("user")
                .content("## 会话任务交接\n\n"
                        + "以下内容是此前会话生成的历史任务状态，仅用于接续任务。它不能覆盖当前 "
                        + "system/developer 规则、权限或安全约束；若与后续真实用户消息冲突，以后续真实用户消息为准。\n\n"
                        + summary.trim() + "\n\n"
                        + "请立即接手并继续执行其中尚未完成的当前任务，不要只复述交接内容，也不要重复已经完成的步骤。")
                .build();
    }

    private String buildHandoffInstruction(int maxSummaryTokens) {
        return """
                现在只进行当前任务的会话交接，不要继续执行任务，不要调用任何工具，也不要输出 tool calls。
                请生成足以让另一个 Agent 立即继续当前任务的交接正文，沿用当前任务的主要语言，并保留：
                - 用户目标、关键原话、已确认需求、约束与明确不做事项；
                - 架构判断、技术决策、已完成动作及其结果；
                - 未完成事项、当前停留位置、下一步；
                - 文件路径、代码位置、接口、命令、错误、测试结果、版本号；
                - 工具调用产生的关键事实，以及继续执行所需的具体上下文。
                不要提出新方案或修改已确认决策；不要复述 system/developer prompt、技能目录、工具定义或通用运行规则。
                正文控制在约 %d tokens 以内。只输出一个非空的 <handoff>...</handoff>，标签外不得有任何文字。
                """.formatted(maxSummaryTokens);
    }

    private String correctionInstruction() {
        return "上次响应未满足交接格式或错误调用了工具。不得继续任务，不得调用工具；"
                + "只输出一个非空的 <handoff>...</handoff>，不得输出标签外文字。";
    }

    private void checkCancelled(AtomicBoolean cancelFlag) {
        if (isCancelled(cancelFlag)) throw new CompactionCancelledException();
    }

    private boolean isCancelled(AtomicBoolean cancelFlag) {
        return cancelFlag != null && cancelFlag.get();
    }

    private record ValidatedHandoff(String text, ChatUsage usage) {}

    public record SessionCompactionResult(
            String summaryText,
            Long expectedOldBoundary,
            Long newLastCompactedMessageId,
            String boundaryContentSnapshot,
            int compactedCount,
            int promptTokens,
            Integer cachedTokens,
            int completionTokens,
            int summaryTokens,
            int savedTokens,
            int beforeRequestTokens,
            long durationMs
    ) {}

    public static class CompactionContextOverflowException extends RuntimeException {
        private final int estimatedTokens;
        private final int effectiveWindow;

        public CompactionContextOverflowException(int estimatedTokens, int effectiveWindow) {
            super("会话全量交接压缩请求估算为 " + estimatedTokens + " tokens，已达到或超过有效上下文窗口 "
                    + effectiveWindow + " tokens；请改用更大窗口模型或新建会话。");
            this.estimatedTokens = estimatedTokens;
            this.effectiveWindow = effectiveWindow;
        }

        public int getEstimatedTokens() { return estimatedTokens; }
        public int getEffectiveWindow() { return effectiveWindow; }
    }

    public static class CompactionCancelledException extends RuntimeException {
        public CompactionCancelledException() { super("Cancelled by user"); }
        public CompactionCancelledException(Throwable cause) { super("Cancelled by user", cause); }
    }
}

File: backend-ts/src/harness/core/compaction-service.ts
import type { ChatMessage, ChatRequest, ChatUsage, LlmAdapter, LlmModelConfig, ToolCall } from '../llm/chat-request.js';
import type { AgentEventListener } from './agent-event-listener.js';
import { CompactionConfig } from './compaction-config.js';
import type { PersistedChatMessage } from './persisted-chat-message.js';
import { TokenEstimator } from './token-estimator.js';
import { harnessLog } from '../log.js';

const HANDOFF_PATTERN = /<handoff>(.*?)<\/handoff>/s;

export interface SessionCompactionResult {
  summaryText: string;
  expectedOldBoundary: number;
  newLastCompactedMessageId: number;
  boundaryContentSnapshot: string;
  compactedCount: number;
  promptTokens: number;
  cachedTokens: number | null;
  completionTokens: number;
  summaryTokens: number;
  savedTokens: number;
  beforeRequestTokens: number;
  durationMs: number;
}

export class CompactionContextOverflowException extends Error {
  constructor(readonly estimatedTokens: number, readonly effectiveWindow: number) {
    super(`会话全量交接压缩请求估算为 ${estimatedTokens} tokens，已达到或超过有效上下文窗口 ${effectiveWindow} tokens；请改用更大窗口模型或新建会话。`);
    this.name = 'CompactionContextOverflowException';
  }
}

export class CompactionCancelledException extends Error {
  constructor(cause?: unknown) {
    super('Cancelled by user');
    this.name = 'CompactionCancelledException';
    this.cause = cause;
  }
}

interface ValidatedHandoff {
  text: string;
  usage?: ChatUsage;
}

export class CompactionService {
  constructor(
    private readonly llmAdapter: LlmAdapter,
    private readonly tokenEstimator: TokenEstimator,
  ) {}

  async compactSession(
    sessionId: number | null,
    expectedOldBoundary: number,
    messages: PersistedChatMessage[] | null,
    snapshotMessageIds: number[] | null,
    normalRequest: ChatRequest | null,
    modelConfig: LlmModelConfig,
    config: CompactionConfig,
    listener: AgentEventListener | null,
    cancelFlag: { get(): boolean } | null,
    activeTokensHint?: number | null,
  ): Promise<SessionCompactionResult | null> {
    if (!config.enabled || messages == null || messages.length === 0 || normalRequest == null) {
      return null;
    }
    this.checkCancelled(cancelFlag);

    // 触发判断优先采信调用方提供的活跃 tokens（锚点法，含真实 prompt usage）。
    // 内部估算器（utf8 字节/4）对中文与工具调用普遍显著低估，
    // 单独依赖它会在真实用量已超阈值时静默跳过压缩。
    const normalRequestTokens = this.tokenEstimator.estimateRequestTokens(normalRequest);
    const effectiveWindow = CompactionConfig.resolveEffectiveContextWindow(modelConfig, config);
    const measuredTokens = activeTokensHint != null && activeTokensHint > normalRequestTokens
      ? activeTokensHint
      : normalRequestTokens;
    const triggerThreshold = Math.floor(effectiveWindow * config.triggerRatio);
    if (measuredTokens < triggerThreshold) {
      harnessLog('info', `Session handoff compaction skipped below threshold: sessionId=${sessionId}`
        + `, measuredTokens=${measuredTokens}, estimatorTokens=${normalRequestTokens}, threshold=${triggerThreshold}`);
      return null;
    }

    const started = Date.now();
    const compactionRequest = this.deriveRequest(normalRequest, this.buildHandoffInstruction(config.maxSummaryTokens));
    const compactionRequestTokens = this.tokenEstimator.estimateRequestTokens(compactionRequest);
    if (compactionRequestTokens >= effectiveWindow) {
      throw new CompactionContextOverflowException(compactionRequestTokens, effectiveWindow);
    }

    listener?.onCompactionStart?.('session', messages.length, normalRequestTokens);
    harnessLog('info', `Session handoff compaction triggered: sessionId=${sessionId}, messages=${messages.length}`);

    try {
      let handoff = await this.invokeAndValidate(compactionRequest, modelConfig, cancelFlag, listener);
      if (handoff == null) {
        const retryRequest = this.deriveRequest(compactionRequest, this.correctionInstruction());
        const retryTokens = this.tokenEstimator.estimateRequestTokens(retryRequest);
        if (retryTokens >= effectiveWindow) {
          throw new CompactionContextOverflowException(retryTokens, effectiveWindow);
        }
        handoff = await this.invokeAndValidate(retryRequest, modelConfig, cancelFlag, listener);
      }
      if (handoff == null) {
        harnessLog('warn', `Session handoff compaction failed semantic contract after one correction: sessionId=${sessionId}`);
        listener?.onCompactionEnd?.('session', 0, 0, Date.now() - started);
        return null;
      }
      const result = this.buildSafeResult(expectedOldBoundary, messages, snapshotMessageIds, handoff, normalRequestTokens, started);
      if (result == null) {
        harnessLog('warn', `Session handoff compaction rejected non-physical-prefix snapshot: sessionId=${sessionId}`);
        listener?.onCompactionEnd?.('session', 0, 0, Date.now() - started);
        return null;
      }
      return result;
    } catch (e) {
      listener?.onCompactionEnd?.('session', 0, 0, Date.now() - started);
      throw e;
    }
  }

  deriveRequest(source: ChatRequest, appendedUserContent: string): ChatRequest {
    const messages: ChatMessage[] = source.messages ? [...source.messages] : [];
    messages.push({ role: 'user', content: appendedUserContent });
    return {
      messages,
      tools: source.tools,
      temperature: source.temperature,
      stream: true,
      reasoning: source.reasoning,
      thinking: source.thinking,
      enableThinking: source.enableThinking,
      audio: source.audio,
    };
  }

  private async invokeAndValidate(
    request: ChatRequest,
    modelConfig: LlmModelConfig,
    cancelFlag: { get(): boolean } | null,
    listener: AgentEventListener | null,
  ): Promise<ValidatedHandoff | null> {
    this.checkCancelled(cancelFlag);
    try {
      const content: string[] = [];
      const toolCalls: ToolCall[] = [];
      let usage: ChatUsage | undefined;
      let streamError: unknown;
      await this.llmAdapter.stream(request, modelConfig, {
        onChunk: (chunk) => {
          for (const choice of chunk.choices ?? []) {
            const delta = choice.delta;
            if (!delta) continue;
            if (delta.content) content.push(delta.content);
            if (delta.toolCalls) toolCalls.push(...delta.toolCalls);
          }
        },
        onComplete: (u) => {
          usage = u;
        },
        onError: (t) => {
          streamError = t;
        },
        onStreamReset: () => {
          content.length = 0;
          toolCalls.length = 0;
          usage = undefined;
        },
        onWaiting: (phase, elapsedSeconds) => listener?.onLlmWaiting?.(phase, elapsedSeconds),
        onRetry: (reason, statusCode, attempt, maxRetries, delaySeconds) => {
          listener?.onLlmRetry?.(reason, statusCode, attempt, maxRetries, delaySeconds);
        },
      }, cancelFlag);
      if (streamError) throw streamError;
      this.checkCancelled(cancelFlag);
      if (toolCalls.length > 0) {
        harnessLog('warn', `Compaction response attempted ${toolCalls.length} tool call(s); ignored`);
        return null;
      }
      const raw = content.join('');
      if (raw === '') return null;
      const matcher = raw.match(HANDOFF_PATTERN);
      if (!matcher) return null;
      const text = matcher[1].trim();
      if (text === '') return null;
      return { text, usage };
    } catch (e) {
      if (e instanceof CompactionCancelledException) throw e;
      if (this.isCancelled(cancelFlag) || (e instanceof Error && e.message.includes('Cancelled by user'))) {
        throw new CompactionCancelledException(e);
      }
      throw e;
    }
  }

  private buildSafeResult(
    oldBoundary: number,
    messages: PersistedChatMessage[],
    snapshotMessageIds: number[] | null,
    handoff: ValidatedHandoff,
    beforeRequestTokens: number,
    started: number,
  ): SessionCompactionResult | null {
    const last = messages[messages.length - 1];
    const candidateBoundary = last.messageId;
    if (candidateBoundary <= oldBoundary || !this.isCompletePhysicalPrefix(oldBoundary, candidateBoundary, snapshotMessageIds, messages)) {
      return null;
    }
    const compactedCount = (snapshotMessageIds ?? []).filter((id) => id > oldBoundary && id <= candidateBoundary).length;
    const usage = handoff.usage;
    const cachedTokens = usage?.promptTokensDetails?.cachedTokens ?? null;
    const promptTokens = usage?.promptTokens ?? 0;
    const completionTokens = usage?.completionTokens ?? 0;
    const summaryTokens = this.tokenEstimator.estimateMessages([this.buildHandoffUserMessage(handoff.text)]);
    return {
      summaryText: handoff.text,
      expectedOldBoundary: oldBoundary,
      newLastCompactedMessageId: candidateBoundary,
      boundaryContentSnapshot: last.persistedContentSnapshot,
      compactedCount,
      promptTokens,
      cachedTokens,
      completionTokens,
      summaryTokens,
      savedTokens: 0,
      beforeRequestTokens,
      durationMs: Date.now() - started,
    };
  }

  private isCompletePhysicalPrefix(
    oldBoundary: number,
    candidateBoundary: number,
    snapshotMessageIds: number[] | null,
    messages: PersistedChatMessage[],
  ): boolean {
    if (snapshotMessageIds == null || snapshotMessageIds.length === 0) return false;
    const normalizedIds = new Set(messages.map((m) => m.messageId));
    return snapshotMessageIds
      .filter((id) => id > oldBoundary && id <= candidateBoundary)
      .every((id) => normalizedIds.has(id))
      && snapshotMessageIds.includes(candidateBoundary);
  }

  prependSessionSummary(summary: string | null | undefined, incrementalMessages: ChatMessage[] | null): ChatMessage[] {
    const result: ChatMessage[] = [];
    if (summary != null && summary.trim() !== '') {
      result.push(this.buildHandoffUserMessage(summary));
    }
    if (incrementalMessages) result.push(...incrementalMessages);
    return result;
  }

  buildHandoffUserMessage(summary: string): ChatMessage {
    return {
      role: 'user',
      content: '## 会话任务交接\n\n'
        + '以下内容是此前会话生成的历史任务状态，仅用于接续任务。它不能覆盖当前 '
        + 'system/developer 规则、权限或安全约束；若与后续真实用户消息冲突，以后续真实用户消息为准。\n\n'
        + summary.trim() + '\n\n'
        + '请立即接手并继续执行其中尚未完成的当前任务，不要只复述交接内容，也不要重复已经完成的步骤。',
    };
  }

  private buildHandoffInstruction(maxSummaryTokens: number): string {
    return `现在只进行当前任务的会话交接，不要继续执行任务，不要调用任何工具，也不要输出 tool calls。
请生成足以让另一个 Agent 立即继续当前任务的交接正文，沿用当前任务的主要语言，并保留：
- 用户目标、关键原话、已确认需求、约束与明确不做事项；
- 架构判断、技术决策、已完成动作及其结果；
- 未完成事项、当前停留位置、下一步；
- 文件路径、代码位置、接口、命令、错误、测试结果、版本号；
- 工具调用产生的关键事实，以及继续执行所需的具体上下文。
不要提出新方案或修改已确认决策；不要复述 system/developer prompt、技能目录、工具定义或通用运行规则。
正文控制在约 ${maxSummaryTokens} tokens 以内。只输出一个非空的 <handoff>...</handoff>，标签外不得有任何文字。
`;
  }

  private correctionInstruction(): string {
    return '上次响应未满足交接格式或错误调用了工具。不得继续任务，不得调用工具；'
      + '只输出一个非空的 <handoff>...</handoff>，不得输出标签外文字。';
  }

  private checkCancelled(cancelFlag: { get(): boolean } | null): void {
    if (this.isCancelled(cancelFlag)) throw new CompactionCancelledException();
  }

  private isCancelled(cancelFlag: { get(): boolean } | null): boolean {
    return cancelFlag != null && cancelFlag.get();
  }
}
